src/Integration/LocalServices/FileTracker.cs

```diff
@@ -0,0 +1,88 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Threading;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.SLCore;
+using SonarLint.VisualStudio.SLCore.Common.Helpers;
+using SonarLint.VisualStudio.SLCore.Common.Models;
+using SonarLint.VisualStudio.SLCore.Core;
+using SonarLint.VisualStudio.SLCore.Service.File;
+using SonarLint.VisualStudio.SLCore.State;
+
+namespace SonarLint.VisualStudio.Integration.LocalServices;
+
+[Export(typeof(IFileTracker))]
+[PartCreationPolicy(CreationPolicy.Shared)]
+public class FileTracker : IFileTracker
+{
+    private readonly ISLCoreServiceProvider serviceProvider;
+    private readonly IActiveConfigScopeTracker activeConfigScopeTracker;
+    private readonly IThreadHandling threadHandling;
+    private readonly IClientFileDtoFactory clientFileDtoFactory;
+    private readonly ILogger logger;
+
+    [ImportingConstructor]
+    public FileTracker(ISLCoreServiceProvider serviceProvider, IActiveConfigScopeTracker activeConfigScopeTracker,
+        IThreadHandling threadHandling, IClientFileDtoFactory clientFileDtoFactory, ILogger logger)
+    {
+        this.serviceProvider = serviceProvider;
+        this.activeConfigScopeTracker = activeConfigScopeTracker;
+        this.threadHandling = threadHandling;
+        this.clientFileDtoFactory = clientFileDtoFactory;
+        this.logger = logger;
+    }
+
+    public void AddFiles(params string[] addedFiles)
+    {
+        threadHandling.RunOnBackgroundThread(() => NotifySlCoreFilesChangedAsync([], addedFiles)).Forget();
+    }
+
+    public void RemoveFiles(params string[] removedFiles)
+    {
+        threadHandling.RunOnBackgroundThread(() => NotifySlCoreFilesChangedAsync(removedFiles, [])).Forget();
+    }
+
+    public void RenameFiles(string[] beforeRenameFiles, string[] afterRenameFiles)
+    {
+        threadHandling.RunOnBackgroundThread(() => NotifySlCoreFilesChangedAsync(beforeRenameFiles, afterRenameFiles))
+            .Forget();
+    }
+
+    private Task NotifySlCoreFilesChangedAsync(string[] removedFiles, string[] addedFiles)
+    {
+        if (serviceProvider.TryGetTransientService(out IFileRpcSLCoreService fileRpcSlCoreService))
+        {
+            var configScope = activeConfigScopeTracker.Current;
+            var clientFiles = addedFiles.Select(fp => clientFileDtoFactory.Create(fp, configScope.Id, configScope.RootPath)).ToList();
+            var removedFileUris = removedFiles.Select(f => new FileUri(f)).ToList();
+
+            fileRpcSlCoreService.DidUpdateFileSystem(new DidUpdateFileSystemParams(
+                removedFileUris, clientFiles));
+        }
+        else
+        {
+            logger.WriteLine("[FileTracker] {0}", SLCoreStrings.ServiceProviderNotInitialized);
```

Comment:
Use `nameof`. Also, I'm wondering if it deserves a method overload (to automate future contextual logs with prefixes). Add a jira ticket